### PR TITLE
Normalize plan before sending to increase the plan apply throughput

### PR DESCRIPTION
**Pull request by arshjohar**
This PR adds normalization of the plan to commit only the diff for stopped and preempted allocs to the raft log to enable better throughput. It also starts using omitempty on some of the structs during msgpack serialization to omit the empty fields.

### CODE REVIEW

File: scheduler/testing.go [hunk 1]
```diff
@@ -130,20 +127,47 @@ func (h *Harness) SubmitPlan(plan *structs.Plan) (*structs.PlanResult, State, er
...
-			Alloc: allocs,
+			Job: plan.Job,
 		},
 		Deployment:        plan.Deployment,
 		DeploymentUpdates: plan.DeploymentUpdates,
 		EvalID:            plan.EvalID,
 		NodePreemptions:   preemptedAllocs,
 	}
 
+	if h.optimizePlan {
```
Comment (arshjohar) [inline scheduler/testing.go]: I didn't find any usages for the SubmitPlan method of Harness, but changed the code to be able to support the newer format of the struct.

File: nomad/structs/structs.go [hunk 2]
```diff
@@ -7253,6 +7265,10 @@ type Allocation struct {
 	ModifyTime int64
 }
 
+// AllocationDiff is a type alias for Allocation used to indicate that a diff is
+// and not the entire allocation
+type AllocationDiff = Allocation
```
Comment (arshjohar) [inline nomad/structs/structs.go]: Using a named type would have been a better idea as an AllocationDiff can't be replaced by Allocation on that case. The downside was that AllocationDiff wouldn't have the same methods defined on the struct, that Allocation does, which probably means rewriting code.

File: nomad/state/state_store.go [hunk 3]
```diff
@@ -223,35 +210,33 @@ func (s *StateStore) UpsertPlanResults(index uint64, results *structs.ApplyPlanR
...
 
-	// Prepare preempted allocs in the plan results for update
-	var preemptedAllocs []*structs.Allocation
-	for _, preemptedAlloc := range results.NodePreemptions {
-		// Look for existing alloc
-		existing, err := txn.First("allocs", "id", preemptedAlloc.ID)
-		if err != nil {
-			return fmt.Errorf("alloc lookup failed: %v", err)
-		}
+	noOfAllocs := len(results.NodePreemptions)
```
Comment (schmichael) [inline nomad/state/state_store.go]: `numAllocs` is probably a more idiomatic name.

File: nomad/state/state_store.go [hunk 4]
```diff
@@ -266,6 +251,28 @@ func (s *StateStore) UpsertPlanResults(index uint64, results *structs.ApplyPlanR
 	return nil
 }
 
+func addComputedAllocAttrs(allocs []*structs.Allocation, job *structs.Job) {
```
Comment (schmichael) [inline nomad/state/state_store.go]: Comment

File: nomad/state/state_store_test.go [hunk 5]
```diff
@@ -140,6 +141,94 @@ func TestStateStore_UpsertPlanResults_AllocationsCreated_Denormalized(t *testing
...
+	}
+	preemptedAlloc := mock.Alloc()
+	preemptedAlloc.Job = job
+	preemptedAllocDiff := &structs.Allocation{
+		ID:                    preemptedAlloc.ID,
+		PreemptedByAllocation: alloc.ID,
+	}
+
+	if err := state.UpsertAllocs(900, []*structs.Allocation{stoppedAlloc, preemptedAlloc}); err != nil {
+		t.Fatalf("err: %v", err)
```
Comment (schmichael) [inline nomad/state/state_store_test.go]: require.NoError(err)

File: nomad/state/state_store_test.go [hunk 6]
```diff
@@ -316,7 +403,7 @@ func TestStateStore_UpsertPlanResults_PreemptedAllocs(t *testing.T) {
 	preempted, err := state.AllocByID(ws, preemptedAlloc.ID)
 	require.NoError(err)
 	require.Equal(preempted.DesiredStatus, structs.AllocDesiredStatusEvict)
-	require.Equal(preempted.DesiredDescription, fmt.Sprintf("Preempted by allocation %v", alloc.ID))
+	require.Equal(preempted.DesiredDescription, fmt.Sprintf("Preempted by alloc ID %v", alloc.ID))
```
Comment (schmichael) [inline nomad/state/state_store_test.go]: Is there a reason we changed this string? Old version seems fine.

File: nomad/structs/structs.go [hunk 7]
```diff
@@ -7139,6 +7148,9 @@ const (
 
 // Allocation is used to allocate the placement of a task group to a node.
 type Allocation struct {
+	// msgpack omit empty fields during serialization
+	_struct bool `codec:",omitempty"` // nolint: structcheck
```
Comment (schmichael) [inline nomad/structs/structs.go]: Is there another way of doing this? Not a big deal, it would just be nice to avoid nolint comments when possible.

Perhaps we should do this globally on our msgpack encoder? I can't imagine why we'd ever want to serialize zero values.

File: nomad/structs/structs.go (hunk 2 above)
Comment (schmichael) [reply nomad/structs/structs.go]: Do we want the same methods defined? Allocation has a lot of methods that will not work or worse: appear to work but not have the intended result.

If there's a small number of methods that need to be shared AllocationDiff I think a named type can be converted easily:

```go
func (ad *AllocationDiff) CopySkipJob() *Allocation {
    return Allocation(ad).CopySkipJob()
}
```

If that's more trouble than it's worth let's remove this type alias, use Allocation directly, and carefully comment the fields that have partial Allocs.

File: nomad/structs/structs.go [hunk 8]
```diff
@@ -8411,6 +8434,27 @@ func (p *Plan) IsNoOp() bool {
 		len(p.DeploymentUpdates) == 0
 }
 
+func (p *Plan) NormalizeAllocations() {
```
Comment (schmichael) [inline nomad/structs/structs.go]: comment

Comment (arshjohar) [reply nomad/structs/structs.go]: @schmichael I'll take a look at this probably in the next week. But, I think in the current shape of things, this isn't possible. I think the flaw is that currently we don't have a separate concept of DTOs, when we send the object over the wire, which is something I did point out in the design discussions. It's not that I don't want to use more strongly typed code, I do and I did explore it. But, switching to DTOs was more work, and we didn't scope it.

File: nomad/structs/structs.go (hunk 7 above)
Comment (arshjohar) [reply nomad/structs/structs.go]: Unfortunately, this is the only way. I remember looking at it, and there's no global option for OmitEmpty.

File: nomad/state/state_store_test.go (hunk 6 above)
Comment (arshjohar) [reply nomad/state/state_store_test.go]: The older version of string in the test was incorrect. The source code has always used "alloc ID". Earlier the test used to setup this data (with the incorrect string), whereas now it is denormalized in the source code and hence, the change.

File: nomad/util.go [hunk 9]
```diff
@@ -145,9 +150,9 @@ func isNomadServer(m serf.Member) (bool, *serverParts) {
 
 // ServersMeetMinimumVersion returns whether the given alive servers are at least on the
 // given Nomad version
-func ServersMeetMinimumVersion(members []serf.Member, minVersion *version.Version) bool {
+func ServersMeetMinimumVersion(members []serf.Member, minVersion *version.Version, checkFailedServers bool) bool {
```
[the remark this answers is not in the document]
Comment (arshjohar) [reply nomad/util.go]: sure

File: nomad/plan_apply_test.go [hunk 10]
```diff
@@ -228,6 +230,154 @@ func TestPlanApply_applyPlan(t *testing.T) {
...
+	plan := &structs.Plan{
+		Job:               alloc.Job,
+		Deployment:        dnew,
+		DeploymentUpdates: updates,
+		EvalID:            eval.ID,
+	}
+
+	// Apply the plan
+	future, err := s1.applyPlan(plan, planRes, snap)
+	assert := assert.New(t)
```
[the remark this answers is not in the document]
Comment (arshjohar) [reply nomad/plan_apply_test.go]: This is mostly following the same pattern as the older denormalized case did (and the same is true for other tests which have a denormalized version from the past). Having said that, I think it's probably not a good idea for only the first failure to matter in a test. The reason being that the programmer doesn't want to go through a cycle wherein they have to go back and forth while fixing a test to uncover all the issues. Especially for a test that tests multiple somewhat independent objects as this one does. In cases where only one object is tested, and that one was nil and there's no point testing it for attributes, it's better in cases like those to stop running the test.

**Comment by DingoEatingFuzz** [comment]
@arshjohar, the `0.9.1-dev` branch has been merged into master. Please reopen this pointed to master.

File: nomad/plan_apply_test.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (preetapan) [reply nomad/plan_apply_test.go]: @arsh - For historic context, we used assert for a while in the past (which is why some old tests do this) and more recently have pretty much standardized on require. In this test for instance, if the plan apply failed and returned an error there's no point in examining allocations expected to be created from that plan. If an alloc look up from the state store errored, the immediate next line checking for some state in the alloc will also fail. More often than not require is what you want anyway, so we've tried to standardize on that for new code.

File: api/allocations.go [hunk 11]
```diff
@@ -86,6 +86,7 @@ type Allocation struct {
 	EvalID                string
 	Name                  string
 	NodeID                string
+	NodeName              string
 	JobID                 string
```
Comment (preetapan) [inline api/allocations.go]: Needs a rebase to master to remove these unrelated changes that have already been merged

File: nomad/plan_apply.go [hunk 12]
```diff
@@ -232,6 +261,39 @@ func (p *planner) applyPlan(plan *structs.Plan, result *structs.PlanResult, snap
 	return future, nil
 }
```
Comment (preetapan) [inline nomad/plan_apply.go]: Pls add comments to all these internal functions for future readability. Comments should make it very clear that the fields set are the only ones the plan applier is responsible for setting and everything else is looked up

File: nomad/state/state_store.go [hunk 13]
```diff
@@ -4100,6 +4107,67 @@ type StateSnapshot struct {
...
+	// Output index for denormalized Allocations
+	j := 0
+
+	for _, allocDiff := range allocDiffs {
+		alloc, err := s.AllocByID(nil, allocDiff.ID)
+		if err != nil {
+			return fmt.Errorf("alloc lookup failed: %v", err)
+		}
+		if alloc == nil {
+			continue
```
Comment (preetapan) [inline nomad/state/state_store.go]: This should return an error that bubbles up such that the plan apply fails and the worker is forced to do a index refresh. 
Otherwise if there's a race between a forced garbage collection and the scheduler making an update to the alloc, the alloc could be gone from the state store before it gets here and silently return true though the update didn't actually make it. 
cc @dadgar to double check the above^

File: nomad/state/state_store_test.go [hunk 14]
```diff
@@ -140,6 +141,94 @@ func TestStateStore_UpsertPlanResults_AllocationsCreated_Denormalized(t *testing
...
+	assert.Equal(preemptedAllocDiff.PreemptedByAllocation, updatedPreemptedAlloc.PreemptedByAllocation)
+	assert.Equal(planModifyIndex, updatedPreemptedAlloc.AllocModifyIndex)
+	assert.Equal(planModifyIndex, updatedPreemptedAlloc.AllocModifyIndex)
+
+	index, err := state.Index("allocs")
+	assert.Nil(err)
+	assert.EqualValues(planModifyIndex, index)
+
+	if watchFired(ws) {
+		t.Fatalf("bad")
```
Comment (preetapan) [inline nomad/state/state_store_test.go]: This was a bad anti pattern in the tests with this error message with no context, lets keep new tests clean. Replace with require or make a better error message.

File: nomad/plan_apply_test.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (arshjohar) [reply nomad/plan_apply_test.go]: My thought process was that plan apply does multiple inserts, and it could fail at any point, so its probably a good idea to know what failed and what didn't. And, doing a nil check for require makes perfect sense, but it's hard to do a require and flip to assert (little value, maintain order of objects asserted). I'll change the test to use require, since that is what the team has standardized on.

File: nomad/state/state_store.go (hunk 13 above)
Comment (dadgar) [reply nomad/state/state_store.go]: I think it is safer to return an error here as it indicates that the scheduler made an update on stale information. I don't think it is likely to ever be hit though because it not being in the state store means that the allocation has been GC'd which is only possible if the user forced GC'd between the scheduler snapshot and plan apply.
